dask_kubernetes/operator/daskcluster.py

```diff
@@ -187,18 +192,79 @@ async def daskcluster_create(spec, name, namespace, logger, **kwargs):
 async def daskworkergroup_create(spec, name, namespace, logger, **kwargs):
     api = kubernetes.client.CoreV1Api()
 
-    scheduler = kubernetes.client.CustomObjectsApi().list_cluster_custom_object(
+    cluster = kubernetes.client.CustomObjectsApi().list_cluster_custom_object(
         group="kubernetes.dask.org", version="v1", plural="daskclusters"
     )
-    scheduler_name = scheduler["items"][0]["metadata"]["name"]
-
-    num_workers = spec["workers"]["replicas"]
+    scheduler_name = cluster["items"][0]["metadata"]["name"]
+    # scheduler_spec = scheduler["items"][0]["spec"]
+    # scheduler_data = build_scheduler_pod_spec(
+    #     name=scheduler_name, image=scheduler_spec.get("image")
+    # )
+    # data = build_worker_group_spec(
+    #     name.split('-')[0],
+    #     spec.get("image"),
+    #     spec.get("replicas"),
+    #     spec.get("resources"),
+    #     spec.get("env"),
+    # )
+    # kopf.adopt(data, owner = cluster["items"][0])
```

Comment:
Having trouble getting the `DaskWorkerGroup` to become adopted by the `DaskCluster`. The `owner` executing without any errors and when I check log the `DaskWorkerGroup` I get this:
```
{
   "apiVersion":"kubernetes.dask.org/v1",
   "kind":"DaskWorkerGroup",
   "metadata":{
      "name":"default-worker-group",
      "ownerReferences":[
         {
            "controller":true,
            "blockOwnerDeletion":true,
            "apiVersion":"kubernetes.dask.org/v1",
            "kind":"DaskCluster",
            "name":"simple-cluster",
            "uid":"a65120a5-ce6f-485f-acbc-513645582d16"
         }
      ],
      "namespace":"default",
      "labels":{
         
      }
   },
   "spec":{
      "image":"daskdev/dask:latest",
      "replicas":3,
      "resources":{
         
      },
      "env":{
         
      }
   }
}
```
Which is what I expect. The `DaskWorkerGroup` resource is now a child of the `DaskCluster`. But when I delete the `DaskCluster` resource, the `DaskWorkerGroup`'s are not deleted. Therefore, the worker pods are not deleted either. It seems like `kopf.adopt` needs to be called inside a handler that pertains to the parent resource eg. `@kopf.oncreate("daskcluster")`. Maybe the `DaskCluster` resource should also hold a list of names of the `DaskWorkerGroup`s as they are added to the cluster. So we would need an update handler. And as `DaskWorkerGroup`s are added, the update handler would adopt them.

Reply:
Maybe the `DaskCluster` resource should also hold a list of names of the `DaskWorkerGroups`. The names would be added to the `DaskCluster` in `daskworkergroup_create` handler. This would update the `DaskCluster`, so we would also need a `daskcluster_update` handler. And as `DaskWorkerGroups` are added, the update handler would adopt them.

Reply:
@jacobtomlinson

Reply:
Hrm that's frustrating. I think we should definitely keep persuing the adoption method rather than having deletion handlers.

Perhaps this is a good time to built a small separate experiment with a couple of CRDs and an operator and try and understand the adoption mechanism better.